persist installation path of extensions

boneskull opened the pull request:
This uses the new v3 version of the manifest schema to add an `installPath` property to every extension in `extensions.yaml`. This fixes an issue where extensions may not be able to be _instantiated_ within an extension dev environment. Previously, Appium would _assume_ the path of an extension is in `node_modules` (relative to `APPIUM_HOME`), but that is not always true. In that respect it's a bit of a hotfix to some changes published in the last `appium` release.

In addition, I've added a small "framework" to migrate schema versions, including their types. Please see `packages/types/lib/manifest/README.md` for more information on how to add a new version of the schema.

boneskull committed:
chore(support): return installation path from npm.installPackage() in addition to package.json

This is a breaking change if someone is actually using the `npm` module out of `@appium/support`.  Personally, I'm gonna go with "no"

The signature of `installPackage()` has changed from returning a parsed `package.json` to returning an `NpmInstallReceipt` object, which contains parsed `package.json` prop `pkg` and `string` prop `installPath`.

boneskull committed:
feat(types): add declarations for versioned manifest files and their types

Contains information on how to increment the schema revision

boneskull committed:
fix(appium): store path to installed extension

This uses the new v3 version of the manifest schema to add an `installPath` property to every extension in `extensions.yaml`.  This fixes an issue where extensions may not be able to be _instantiated_ within an extension dev environment.

boneskull committed:
feat(appium): implement schema migrations

This adds a module `lib/extension/manifest-migrations` which contains functions that migrate `extension.yaml` files to new versions.

- constant `CONFIG_SCHEMA_REV` renamed to `CURRENT_SCHEMA_REV` and moved into `constants` module
- removes references to removed `PluginRecord` and `DriverRecord` types
- `ExtensionFields` is now `ExtInstallReceipt`
- renamed private fields in `Manifest`; see L365-385 for logic changes
- updated test code

boneskull committed:
chore(): stop streaming lerna output

this will make test output easier to read

mykola-mokhnach commented on packages/appium/lib/extension/manifest-migrations.js:
```diff
@@ -0,0 +1,94 @@
+   * @type {Migration}
+   */
+  3: (data) => {
+    let shouldSync = false;
```
do we need this intermediate variable at all?

boneskull replied:
not technically, no, but it's just for clarity of intent

mykola-mokhnach commented on packages/appium/lib/extension/manifest.js:
```diff
@@ -419,34 +425,35 @@ export class Manifest {
-            `Please ensure it is writable. Original error: ${err.message}`
+          `Appium could not write to manifest at ${this.#manifestPath} using APPIUM_HOME ${
+            this.#appiumHome
+          }. ` + `Please ensure it is writable. Original error: ${err.message}`
```
extra catenation here

mykola-mokhnach commented on packages/appium/test/fixtures/manifest/v3.yaml:
```diff
@@ -0,0 +1,27 @@
+    pkgName: '@appium/fake-plugin'
+    version: 1.2.2
+    installType: local
+    installSpec: /Users/alice/projects/appium/node_modules/@appium/fake-plugin
```
I thought installSpec should only have ext name...

mykola-mokhnach commented on packages/appium/lib/extension/manifest-migrations.js:
```diff
@@ -0,0 +1,94 @@
+   *
+   * @type {Migration}
+   */
+  3: (data) => {
```
shall we use the constant instead?

boneskull replied:
There is no constant for this...

There's `CURRENT_SCHEMA_REV = 3`, but that is supposed to be updated when we increment the revision, so we cannot use it.

However: It's probably a good idea to define one anyway, so I'll do that.

boneskull replied:
OK, this is done and I added a note in the `README` about it

KazuCocoa reviewed (approved):
lgtm

jlipps commented on packages/appium/types/manifest/README.md:
```diff
@@ -0,0 +1,29 @@
+
+## How To Add A New Version
+
+The manifest file format composed of types used elsewhere in our codebase.  This is probably a design flaw, but the fact remains that the manifest file's format is strongly coupled to Appium's internal logic around extensions.
```
`is composed of`

boneskull committed:
fix(appium): fix problematic ExtManifestWithSchema type

This type was having trouble due to a potential bug in `type-fest`

Also fixed an issue with an unused variable
